components/authz-service/storage/v2/postgres/postgres.go

```diff
@@ -1267,6 +1262,105 @@ func (p *pg) ListRulesForProject(ctx context.Context, projectID string) ([]*v2.R
 	return rules, nil
 }
 
+// ApplyStagedRules begins a db transaction, locks the rule tables, moves all staged rule updates
+// and deletes into the applied rule table, and returns the database transaction. The transaction is returned
+// so that other non-database concerns can be completed before freeing the lock to avoid race conditions.
+func (p *pg) ApplyStagedRules(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	tx, err := p.db.BeginTx(ctx, nil)
+	if err != nil {
+		return p.processError(err)
+	}
+
+	_, err = tx.ExecContext(ctx,
+		`LOCK TABLE iam_project_rules; LOCK TABLE iam_rule_conditions;
+			LOCK TABLE iam_staged_project_rules; LOCK TABLE iam_staged_rule_conditions; `,
+	)
+	if err != nil {
+		return p.processError(err)
+	}
+
+	// Upsert all staged rules into applied rules marked for update, returning the id and db_id
+	// of all rules affected so we can update their conditions below.
+	rows, err := tx.QueryContext(ctx,
+		`INSERT INTO iam_project_rules (id, project_id, name, type)
+				SELECT s.id, s.project_id, s.name, s.type
+					FROM iam_staged_project_rules AS s
+					WHERE deleted=false
+				ON CONFLICT (id) DO UPDATE
+				SET name=excluded.name, type=excluded.type
+				RETURNING id, db_id;`)
+	defer func() {
+		if err := rows.Close(); err != nil {
```

Comment:
I believe this needs to go below the `err != nil`; if `err == nil` in line 1301, the defer here would run and call `Close()` on `rows == nil`